_posts/2017-09-26-testing-a-react-app.md

```diff
@@ -0,0 +1,131 @@
+The first thing I decided to test was checking that the component renders without failures. This is how you can do it with Jest:
+```
+    import ReactDOM from 'react-dom';
+
+    describe('Counter component', () => {
+        test('renders without crashing', () => {
+            const div = document.createElement('div');
+            expect(() => ReactDOM.render(<Counter/>, div)).not.toThrow();
+        });
+    });
+```
+This test renders the Counter component in the DOM and checks that no exception is thrown. It doesn't test much but if your component has some child dependencies that are broken or you made a change that breaks the DOM it will tell you.
+It is also useful when you migrate or upgrade major versions of certain libraries (e.g. React) and want to check that it does not have major impacts such as breaking your DOM.
+
+
+#### Testing DOM content of a Component
+
+Next, I tested that the actual DOM (the HTML tree) was correct when rendered for the first time. I did that using Jest [snapshot functionality](https://facebook.github.io/jest/docs/en/snapshot-testing.html): it basically allows you to check that the DOM generated for your component hasn't changed by comparing the HTML tree against a snapshot you are providing to the test. You can provide the snapshot yourself or you can use the one created automatically the first time you run the test (just make sure it does not contain an error already).
+
+This is my test:
+```
+    import testRenderer from 'react-test-renderer';
+    import Counter from './Counter';
+
+    describe('Counter component', () => {
+        test('DOM matches the snapshot', () => {
+            const component = testRenderer.create(
+                <Counter/>
+            );
+            let tree = component.toJSON();
+            expect(tree).toMatchSnapshot("rendering");
+        });
+    });
+```
+And this is the snapshot that was generated the first time I run the test:
+```
+    // Jest Snapshot v1, https://goo.gl/fbAQLP
+
+    exports[`rendering 1`] = `
+    <button
+    onClick={[Function]}
+    >
+    0
+    </button>
+    `;
+```
+This is the first time I do that kind of testing and I wonder in practice how useful and maintanable this is. If you add more cases or if your component is more complex how developer-friendly does it become? I need to have used it on a real life project to decide but for now I tested the functionality and I found it pretty cool for a simple component like mine (maybe this is the key, it forces you to keep your component super simple ;))
+
+Note that for this test I also added an extra npm dependency to [react-test-renderer](https://www.npmjs.com/package/react-test-renderer): `npm install --save-dev react-test-renderer` (react-test-renderer is a utility that provides renderers that don't depend on DOM).
+
```

Comment:
Maybe it would be cool to have a dependency summary somewhere in the article (at the end?)

Reply:
done!